File: app/src/main/java/com/arcao/geocaching4locus/base/util/DownloadingUtil.java
package com.arcao.geocaching4locus.base.util;

import com.arcao.geocaching4locus.base.constants.AppConstants;

/**
 * Created by Arcao on 12.02.2018.
 */
public final class DownloadingUtil {
    private DownloadingUtil() {
    }

    public static int computeItemsPerRequest(int currentItemsPerRequest, long startTimeMillis) {
        int itemsPerRequest = currentItemsPerRequest;
        long requestDuration = System.currentTimeMillis() - startTimeMillis;

        // keep the request time between ADAPTIVE_DOWNLOADING_MIN_TIME_MS and ADAPTIVE_DOWNLOADING_MAX_TIME_MS
        if (requestDuration < AppConstants.ADAPTIVE_DOWNLOADING_MIN_TIME_MS)
            itemsPerRequest += AppConstants.ADAPTIVE_DOWNLOADING_STEP;

        if (requestDuration > AppConstants.ADAPTIVE_DOWNLOADING_MAX_TIME_MS)
            itemsPerRequest -= AppConstants.ADAPTIVE_DOWNLOADING_STEP;

        // keep the value in a range
        itemsPerRequest = Math.max(itemsPerRequest, AppConstants.ADAPTIVE_DOWNLOADING_MIN_ITEMS);
        itemsPerRequest = Math.min(itemsPerRequest, AppConstants.ADAPTIVE_DOWNLOADING_MAX_ITEMS);

        return itemsPerRequest;
    }
}

File: app/src/main/java/com/arcao/geocaching4locus/base/util/DownloadingUtil.kt
package com.arcao.geocaching4locus.base.util

import kotlin.math.max
import kotlin.math.min

object DownloadingUtil {
    /* Adaptive downloading configuration */
    const val MIN_REQUEST_SIZE = 10
    const val REQUEST_SIZE_INCREMENT = 20
    const val REQUEST_SIZE_DECREMENT = REQUEST_SIZE_INCREMENT
    const val MIN_REQUEST_TIME_MS =
        3500 // more than time required for 30 calls per minute
    const val MAX_REQUEST_TIME_MS = 10000

    fun computeRequestSize(
        currentRequestSize: Int,
        maxRequestSize: Int,
        startTimeMillis: Long
    ): Int {
        var requestSize = currentRequestSize
        val requestDuration = System.currentTimeMillis() - startTimeMillis

        // keep the request time between MIN_REQUEST_TIME_MS and MAX_REQUEST_TIME_MS
        if (requestDuration < MIN_REQUEST_TIME_MS)
            requestSize += REQUEST_SIZE_INCREMENT

        if (requestDuration > MAX_REQUEST_TIME_MS)
            requestSize -= REQUEST_SIZE_DECREMENT

        // keep the value in a range
        requestSize = max(requestSize, MIN_REQUEST_SIZE)
        requestSize = min(requestSize, maxRequestSize)

        return requestSize
    }
}
